solidity/contracts/bank/Bank.sol

```diff
@@ -0,0 +1,299 @@
+    ///         construct EIP2612 signature provided to `permit` function.
+    mapping(address => uint256) public nonce;
+
+    uint256 public immutable cachedChainId;
+    bytes32 public immutable cachedDomainSeparator;
+
+    /// @notice Returns EIP2612 Permit message hash. Used to construct EIP2612
+    ///         signature provided to `permit` function.
+    bytes32 public constant PERMIT_TYPEHASH =
+        keccak256(
+            "Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)"
+        );
+
+    event BalanceTransferred(
+        address indexed from,
+        address indexed to,
+        uint256 amount
+    );
+
+    event BalanceApproved(
+        address indexed owner,
+        address indexed spender,
+        uint256 amount
+    );
+
+    event BalanceIncreased(address indexed owner, uint256 amount);
+
+    event BalanceDecreased(address indexed owner, uint256 amount);
+
+    event BridgeUpdated(address newBridge);
+
+    modifier onlyBridge() {
+        require(msg.sender == address(bridge), "Caller is not the bridge");
+        _;
+    }
+
+    constructor() {
+        cachedChainId = block.chainid;
+        cachedDomainSeparator = buildDomainSeparator();
+    }
+
+    /// @notice Allows the Governance to upgrade the Bridge address.
+    /// @dev The function does not implement any governance delay and does not
+    ///      check the status of the Bridge. The Governance implementation needs
+    ///      to ensure all requirements for the upgrade are satisfied before
+    ///      executing this function.
+    function updateBridge(address _bridge) external onlyOwner {
+        require(_bridge != address(0), "Bridge address must not be 0x0");
+        bridge = _bridge;
+        emit BridgeUpdated(_bridge);
+    }
+
+    /// @notice Moves the given `amount` of balance from the caller to
+    ///         `recipient`.
+    /// @dev Requirements:
+    ///       - `recipient` cannot be the zero address,
+    ///       - the caller must have a balance of at least `amount`.
+    function transferBalance(address recipient, uint256 amount) external {
+        _transferBalance(msg.sender, recipient, amount);
+    }
+
+    /// @notice Sets `amount` as the allowance of `spender` over the caller's
+    ///         balance.
+    /// @dev If the `amount` is set to `type(uint256).max` then
+    ///      `transferBalanceFrom` will not reduce an allowance.
+    ///      Beware that changing an allowance with this function brings the
+    ///      risk that someone may use both the old and the new allowance by
+    ///      unfortunate transaction ordering. One possible solution to mitigate
+    ///      this race condition is to first reduce the spender's allowance to 0
+    ///      and set the desired value afterwards:
+    ///      https://github.com/ethereum/EIPs/issues/20#issuecomment-263524729
+    function approveBalance(address spender, uint256 amount) external {
```

Comment:
We don't add `increaseAllowance` and `decreaseAllowance` methods that can mitigate the problem described in the comment. Is there any particular reason for not adding them here?

Reply:
Good catch. No particular reason, added in https://github.com/keep-network/tbtc-v2/pull/84/commits/0e15708c0dfa651fac9af15a64652c3211b123da.